Speeding up docs build in CI using caching.

dpeachpeach opened the pull request:
FOR ISSUE #32687 

**Description**

This speeds up the 'Docs Build' step of the docs build job by caching the results of building the docs from prior actions. 

This kind of 'skips' the step 'build docs' altogether (in the successful case where there is a cache hit) and doesn't really provide any improvement to speeding up building the docs in case they were modified.

 I didn't know if that's what you guys were going for but I thought it could be as a decent amount of the contributions made don't affect building the docs directly. In those cases it speeds up the action by eliminating building the docs entirely. (Drawback here is another element present in the cache fields.)

I was also thinking of other approaches to speed things up. If this isn't sufficient / what you're looking for I can find a different approach and improve it further! Give me any feedback you have I'll go back to work on this.

**Verification**

I tested the action on my own fork, which showed that it cached correctly and provided an elimination of the 'build docs' step in case of cache hit. It might take two runs to store the cache properly but barring that it seems to be a success.

dpeachpeach committed:
Implement Caching on 'Build docs' step

dpeachpeach committed:
Fixing syntax errors in prior edit.

dpeachpeach committed:
Fixing another syntax error.

dpeachpeach committed:
Fixed path error.

Now that docs were built earlier, should cache correctly though.

adeebshihadeh commented:
This is close but not quite right.

We want to build able to incrementally build using the cache, like we can with our scons build cache.

dpeachpeach committed:
Merge branch 'commaai:master' into master

dpeachpeach committed:
Putting Parallelism into Sphinx-build

Doubtful this will make a change, as parallelism already accounted for in the github action. Let's see how it goes anyway!

dpeachpeach committed:
Commenting out action-level caching

Although this speeds up times in some cases, maintainers prefer caching done at the sphinx-build level.

dpeachpeach committed:
Merge pull request #1 from dpeachpeach/improving_ci_times

Checking to see how CI times are improved from sphinx-build level parallelism

dpeachpeach committed:
Squashed commits

dpeachpeach committed:
Merge PR: Reducing CI times

Removing the build directory on every make wiped the cache sphinx implements by itself. This reverts that behavior and also performs some refactoring.

dpeachpeach committed:
Merge branch 'commaai:master' into master

dpeachpeach commented:
I took your feedback and tried to improve it further. I think I've got it pending your review but I just wanted to provide some explanation behind the edits + broad strokes bullet points:

- It turns out Sphinx already implements caching implicitly! The reason why it didn't manifest itself in building the docs for comma was that the build directory would be deleted on every build. That is where the automatically generated caches (or pickles) (.doctrees) were stored. I stopped that behavior and now it runs much faster.
- I also added in parallelism at the sphinx-build call, but when I had this undergo testing on my fork it didn't change anything about the performance (probably because parallelism is already done through the github action). This can be removed to clean up the code.
- I went through the route of looking at some optimizations that could affect the theme but I figured those were design decisions and I didn't want to intrude.
- I changed the paths of some of the directory alias definitions to more closely follow definitions I saw in the docs.

When undergoing testing on my fork, this seems to have resulted in the same results with faster compilation time. 
Let me know if you have any more feedback!

dpeachpeach committed:
Deleted action-level caching

dpeachpeach committed:
Clearing out fluff

adeebshihadeh commented:
Sounds good, I'll review once you mark it as ready for review.

> * I went through the route of looking at some optimizations that could affect the theme but I figured those were design decisions and I didn't want to intrude.

We basically got it setup and haven't touched it since, so design changes are welcome!

adeebshihadeh commented:
I see a whole bunch of warnings and errors; are those not new?

https://github.com/commaai/openpilot/actions/runs/9589108324/job/26442355846?pr=32766

dpeachpeach commented:
They seem to be the same warnings / errors that were present in the sample you showed in the issue:
https://github.com/commaai/openpilot/actions/runs/9454981147/job/26043750673

However I was worried about that. I was going to try to debug a little bit to see if any of those were related to my implementation before reviewing.

dpeachpeach commented:
Just found a serious issue @adeebshihadeh, going to have to work on this a bit more. Sorry for prematurely setting up the PR!

adeebshihadeh commented:
No worries, take your time :)

You can also share your new docs on Discord and ask people to poke around for bugs

dpeachpeach committed:
Removed line sabotaging search.

dpeachpeach committed:
Merge pull request #3 from dpeachpeach/improving-ci-speeds

Removed line sabotaging search.

dpeachpeach commented:
All right, fixed the issue! When I test the generated html through local hosting, everything seems equivalent. (Compared it side by side with the docs available online). I'll make a post on the discord for anyone to take a look as well but otherwise I think this is ready!

adeebshihadeh commented on docs/Makefile:
```diff
@@ -25,19 +25,20 @@ clean:
 
 	@echo "Copying docs & config to build folder..."
-	cp -a "$(DOCSDIR)" "$(BUILDDIR)"
+	# cp -a "$(DOCSDIR)" "$(BUILDDIR)"
```
this should either be added back or removed

dpeachpeach replied:
Removed it

adeebshihadeh reviewed (commented):
It looks like this now fixes the docs build system to build incrementally, but I don't see how it caches in CI to take advantage of that?

dpeachpeach committed:
Removing commented out line.

Forgot to do this earlier.

dpeachpeach commented:
> It looks like this now fixes the docs build system to build incrementally, but I don't see how it caches in CI to take advantage of that?

Initially, when I started this PR I thought the solution to speeding up 'build docs' lay somewhere in the github action. Once you mentioned incrementally building the cache I gave up speeding anything up through modifying the CI implementation and pivoted to trying to fix sphinx-build's behavior. (And why it wasn't retrieving its cache for incremental builds) 

It seems like fixing sphinx-build has enabled the 'build docs' step to reach the target time (even though I don't do any explicit caching in the CI) because sphinx automatically builds a cache in the form of the doctrees directory, which is present in the build directory that the github action outputs. Now that we're not removing the build directory on every iteration of running the action, we get the CI-level caching for free.

A lot of this is stuff I've learned in the past day or so, but I believe it to be correct. I'm no expert so please provide any feedback / correct me if I'm completely wrong.
